### PR TITLE
feat: add depth flag

### PR DESCRIPTION
**Pull request by yordis**

### CODE REVIEW

**Comment by yordis** [comment]
@crenshaw-dev could help me to get to the finish line with this one?

**Comment by crenshaw-dev** [comment]
@yordis the code looks to be in good shape. We probably need some docs.

I'd also really like to see how this behaves on a large monorepo over time. I am curious if the shallow fetch is going to cause inefficient storage and slowly fill up the repo server.

**Comment by yordis** [comment]
> I'd also really like to see how this behaves on a large, active monorepo over time. I am curious if the shallow fetch is going to cause inefficient storage and slowly fill up the repo server.

We should synchronize communication with the community, it should, in theory, be faster download and cheaper storage ...

> We probably need some docs.

Where should I add it?

**Comment by blakepettersson** [comment]
>  We should synchronize communication with the community, it should, in theory, be faster download

AFAIU the issue is that it's more expensive with shallow clones (see #5467 and its linked issues)

**Comment by yordis** [comment]
Hey peeps, what is the resolution here?

**Comment by crenshaw-dev** [comment]
I think documentation should be added both of these places:

1. https://argo-cd.readthedocs.io/en/latest/operator-manual/high_availability/#monorepo-scaling-considerations
2. https://argo-cd.readthedocs.io/en/latest/operator-manual/declarative-setup/#repositories

> AFAIU the issue is that it's more expensive with shallow clones (see https://github.com/argoproj/argo-cd/issues/5467 and its linked issues)

Yep. I'm not a git expert, but I could definitely imagine "shallow clone" meaning "fresh copy stored with every pull." A fresh copy might not be as light to download or as easy to store as the diff between the current commit and the previous commit.

In other words, we need someone to run this code on a large, frequently-updated monorepo for at least a few days and compare repo-server behavior to the same monorepo fetched without `depth: 1`.

**Comment by yordis** [comment]
@crenshaw-dev do you mind giving me a hand with the docs? Honestly I am getting stuck with it, I would appreciate if you can help me with it and I get to learn from you.

**Comment by blakepettersson** [comment]
What is the status of #14272 btw? It seems as if this would be a preferable solution (in most cases)

**Comment by yordis** [comment]
I am gonna go back to #14272 since they are some caveats I need to make sure it works:

1. The sparse checkout sticks around
2. Ideally we could use worktree in order to keep the concurrency and avoid conflict between different spare checkout

So, I knew how to get the `depth` in based on what I learned from that work, so I switched to add this feature first, so @crenshaw-dev could gather info about its usage and buy a bit of more time** (assuming here)

**Comment by todaywasawesome** [comment]
> What is the status of #14272 btw? It seems as if this would be a preferable solution (in most cases)

I think they each have their use case. Honestly curious to use them together for a repo that I have.

**Comment by todaywasawesome** [comment]
@yordis We can test on this repo. It's not a super *large* repo but it has 10k commits so the shallow clone is ideal. https://github.com/aperture-sci/DevOps

It's also updated several times a day.

**Comment by yordis** [comment]
@todaywasawesome any support to validate the feature will help us, especially if it is being done in close to production usage

**Comment by yordis** [comment]
Hey folks, are you still interested in taking this feature?

From what I can tell, it will be a matter of time before people ask to pass this flag due to the nature of wrapping `git.`

But it is not up to me; I would appreciate any resolution regarding what needs to happen with this PR.

**Comment by crenshaw-dev** [comment]
> In other words, we need someone to run this code on a large, frequently-updated monorepo for at least a few days and compare repo-server behavior to the same monorepo fetched without depth: 1.

I think this is the next step towards getting this merged. Until we can confirm that setting the depth flag doesn't blow up storage due to inefficient local repo maintenance, I don't think we should merge it.

**Comment by todaywasawesome** [comment]
@yordis I'm setting some time for myself to come back and work on this is the next few weeks.

**Comment by yordis** [comment]
I rebased, hopefully still working fine

**Comment by batazor** [comment]
any news?

**Comment by yordis** [comment]
🤷🏻

**Comment by crenshaw-dev** [comment]
I think we still need evidence that enabling this doesn't blow up storage for a given repo over time.

**Comment by crenshaw-dev** [comment]
Putting this here for reference from other places:

Sparse and shallow git repos are common requests for monorepos. People are accustomed to using these features in CI pipelines. But these settings are much easier to get right in CI pipelines because _you throw away the clone when you're done with it_. Argo CD maintains a persistent clone on the repo-server, allowing concurrent access to the same clone until the repo-server restarts. When managing a persistent clone, you have to handle cases which would be safe to ignore for a throw-away clone.

For example:

1) What happens if two applications with different depths/paths access the same repo at the same time?
2) If I change one of these settings, when does it reflect? Immediately? On next checkout?
3) Is storage efficiency impacted? i.e. is the size of my clone going to blow up over time?
4) Is CPU use impacted? Will I see CPU spikes due to cleanup processes?
5) Is there a need for manual cleanups? When do you call them? Are there concurrency concerns when calling them?

The concurrency concerns are different depending on whether you configure depth/paths at the app level or the repo level. If at the repo level, you're less likely to encounter races, but it's still possible.

**Comment by crenshaw-dev** [comment]
Specific to this PR: does any depth besides "1" ever make sense for Argo CD's use case?

**Comment by crenshaw-dev** [comment]
Consolidating conversation here: https://github.com/argoproj/argo-cd/issues/11198
